src/lib/mina_graphql/mina_graphql.ml

```diff
@@ -4292,85 +4261,110 @@ module Queries = struct
-            let open Deferred.Result.Let_syntax in
-            let block = Transition_frontier.Breadcrumb.(block best_tip) in
-            let blockchain_length = Mina_block.blockchain_length block in
-            let global_slot =
-              Mina_block.consensus_state block
-              |> Consensus.Data.Consensus_state.curr_global_slot
-            in
-            let staged_ledger =
-              Transition_frontier.Breadcrumb.staged_ledger best_tip
-              |> Staged_ledger.ledger
-            in
-            let protocol_state =
-              Transition_frontier.Breadcrumb.protocol_state best_tip
-            in
-            let consensus =
-              Mina_state.Protocol_state.consensus_state protocol_state
-            in
-            let staking_epoch =
-              Consensus.Proof_of_stake.Data.Consensus_state.staking_epoch_data
-                consensus
-            in
-            let next_epoch =
-              Consensus.Proof_of_stake.Data.Consensus_state.next_epoch_data
-                consensus
-            in
-            let staking_epoch_seed =
-              Mina_base.Epoch_seed.to_base58_check
-                staking_epoch.Mina_base.Epoch_data.Poly.seed
-            in
-            let next_epoch_seed =
-              Mina_base.Epoch_seed.to_base58_check
-                next_epoch.Mina_base.Epoch_data.Poly.seed
-            in
-            let runtime_config = Mina_lib.runtime_config mina in
-            let%bind staking_ledger =
-              match Mina_lib.staking_ledger mina with
-              | None ->
-                  Deferred.Result.fail "Staking ledger is not initialized."
-              | Some (Genesis_epoch_ledger l) ->
-                  return (Ledger.Any_ledger.cast (module Ledger) l)
-              | Some (Ledger_db l) ->
-                  return (Ledger.Any_ledger.cast (module Ledger.Db) l)
-            in
+      ~args:
+        Arg.
+          [ arg "stateHash" ~doc:"The state hash of the desired block"
+              ~typ:string
+          ; arg "height"
+              ~doc:"The height of the desired block in the best chain" ~typ:int
+          ]
+      ~resolve:(fun { ctx = mina; _ } () state_hash_opt block_height_opt ->
+        let open Deferred.Result.Let_syntax in
+        let%bind breadcrumb =
+          match (state_hash_opt, block_height_opt) with
+          | None, None -> (
+              match Mina_lib.best_tip mina with
+              | `Bootstrapping ->
+                  Deferred.Result.fail "Daemon is bootstrapping"
+              | `Active breadcrumb ->
+                  return breadcrumb )
+          | Some state_hash_base58, None ->
+              let%bind state_hash =
+                State_hash.of_base58_check state_hash_base58
+                |> Result.map_error ~f:Error.to_string_hum
+                |> Deferred.return
+              in
+              Mina_lib.best_chain_block_by_state_hash mina state_hash
+              |> Deferred.return
```

Comment:
This might be cleaner as a pipeline, using `>>=` in place of the bind.

Reply:
Not sure if that's really cleaner, but here you are: 6f44a34e98.